File: assignment 5/python/programs/4.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class SinglyCircularLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the beginning of the circular linked list
    def insert_at_beginning(self, data):
        new_node = Node(data)
        if self.head is None:
            new_node.next = new_node
            self.head = new_node
        else:
            current = self.head
            while current.next != self.head:
                current = current.next
            new_node.next = self.head
            current.next = new_node
            self.head = new_node

    # Insert a node at the end of the circular linked list
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            new_node.next = new_node
            self.head = new_node
        else:
            current = self.head
            while current.next != self.head:
                current = current.next
            new_node.next = self.head
            current.next = new_node

    # Insert a node after a specific element
    def insert_after(self, data, key):
        new_node = Node(data)
        if self.head is None:
            print("List is empty. Cannot perform insertion.")
            return
        current = self.head
        while current:
            if current.data == key:
                new_node.next = current.next
                current.next = new_node
                return
            current = current.next
            if current == self.head:
                print(f"Element {key} not found in the list.")
                return
        print(f"Element {key} not found in the list.")

    # Delete a node by value
    def delete(self, data):
        if self.head is None:
            print("List is empty. Cannot perform deletion.")
            return
        current = self.head
        prev = None
        while True:
            if current.data == data:
                if current.next == self.head:
                    if prev:
                        prev.next = self.head
                    else:
                        self.head = None
                else:
                    if prev:
                        prev.next = current.next
                    else:
                        temp = self.head
                        while temp.next != self.head:
                            temp = temp.next
                        self.head = current.next
                        temp.next = self.head
                return
            prev = current
            current = current.next
            if current == self.head:
                print(f"Element {data} not found in the list.")
                return

    # Display the circular linked list
    def display(self):
        if self.head is None:
            print("Circular Linked List is empty.")
            return

        current = self.head
        while True:
            print(current.data, end=" -> ")
            current = current.next
            if current == self.head:
                break
        print("Head")


if __name__ == "__main__":
    circular_linked_list = SinglyCircularLinkedList()

    # Insert elements into the circular linked list
    circular_linked_list.insert_at_end(1)
    circular_linked_list.insert_at_end(2)
    circular_linked_list.insert_at_end(3)

    # Display the circular linked list
    print("Circular Linked List:")
    circular_linked_list.display()

    # Insert an element after a specific element
    circular_linked_list.insert_after(4, 2)

    # Display the circular linked list after insertion
    print("Circular Linked List after Insertion:")
    circular_linked_list.display()

    # Delete an element from the circular linked list
    circular_linked_list.delete(2)

    # Display the circular linked list after deletion
    print("Circular Linked List after Deletion:")
    circular_linked_list.display()

File: assignment 5/c/programs/4.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for a node in the circular linked list
struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("Memory allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the circular linked list
struct Node* insertAtEnd(struct Node* head, int data) {
    struct Node* newNode = createNode(data);
    if (head == NULL) {
        newNode->next = newNode; // Circular link to itself
        return newNode;
    }

    struct Node* last = head;
    while (last->next != head) {
        last = last->next;
    }

    newNode->next = head;
    last->next = newNode;
    return head;
}

// Function to delete a node with a specific data value from the circular linked list
struct Node* deleteElement(struct Node* head, int key) {
    if (head == NULL) {
        printf("List is empty.\n");
        return NULL;
    }

    struct Node* current = head;
    struct Node* previous = NULL;

    // Find the node to be deleted
    do {
        if (current->data == key) {
            if (previous == NULL) {
                // If the node to be deleted is the head node
                struct Node* last = head;
                while (last->next != head) {
                    last = last->next;
                }
                head = head->next;
                last->next = head;
                free(current);
                return head;
            } else {
                previous->next = current->next;
                free(current);
                return head;
            }
        }
        previous = current;
        current = current->next;
    } while (current != head);

    printf("Element not found in the list.\n");
    return head;
}

// Function to display the circular linked list
void display(struct Node* head) {
    if (head == NULL) {
        printf("List is empty.\n");
        return;
    }

    struct Node* current = head;
    do {
        printf("%d -> ", current->data);
        current = current->next;
    } while (current != head);

    printf("\n");
}

int main() {
    struct Node* head = NULL;

    // Initialize a predefined circular linked list
    head = insertAtEnd(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtEnd(head, 30);

    printf("Initial Circular Linked List: ");
    display(head);

    int data;

    // First time for insertion
    printf("\nEnter data to insert: ");
    scanf("%d", &data);
    head = insertAtEnd(head, data);
    printf("Updated Circular Linked List: ");
    display(head);

    // Second time for deletion
    printf("\nEnter data to delete: ");
    scanf("%d", &data);
    head = deleteElement(head, data);
    printf("Updated Circular Linked List: ");
    display(head);

    return 0;
}
